libraries/chain/fork_database.cpp

```diff
@@ -56,8 +56,8 @@ namespace eosio::chain {
 
    // match comparison of by_best_branch
    bool first_preferred( const block_state& lhs, const block_state& rhs ) {
-      return std::make_tuple(lhs.last_final_block_num(), lhs.last_qc_block_num(), lhs.timestamp()) >
-             std::make_tuple(rhs.last_final_block_num(), rhs.last_qc_block_num(), rhs.timestamp());
+      return std::make_tuple(lhs.last_final_block_num(), lhs.latest_qc_block_num(), lhs.timestamp()) >
+             std::make_tuple(rhs.last_final_block_num(), rhs.latest_qc_block_num(), rhs.timestamp());
```

Comment:
Why is this still using `last_final_block_num()` and `latest_qc_block_num()` when the index for the `by_best_branch` tag has changed?

Reply:
OK, thanks for the catch, fixing it and also [log_fork_comparison](https://github.com/AntelopeIO/spring/blob/e9a3c4ae6b6d89d702a1763a651ddb0eb8133065/libraries/chain/fork_database.cpp#L35-L42).

I wonder why `first_preferred` doesn't have the `id` as a final criteria. I think we should add it (switching to `sha256_greater` for the `by_best_branch` type.

Reply:
After I updated `first_preferred to use `timestamp`, I hit  an assert in `finality_core`:

![image](https://github.com/AntelopeIO/spring/assets/1496907/4be86e9b-b722-4f52-b796-3cc3658708c4)

This happens when calling `mark_valid` at line 129 below:

https://github.com/AntelopeIO/spring/blob/b48c9349a40349a16512fbe74575aceeb478dfe9/unittests/fork_db_tests.cpp#L127-L136

`first_preferred` calls `last_final_block_timestamp`, which calles `get_block_reference`:

https://github.com/AntelopeIO/spring/blob/b48c9349a40349a16512fbe74575aceeb478dfe9/libraries/chain/finality/finality_core.cpp#L69-L72

and the assert fails in  `get_block_reference`, both `block_num` and `current_block_num()` values are 10.

Reply:
Actually, the `refs` vector is empty. This is most likely a test issue. The `block_state` are manually constructed and no qc_claim is received to advance finality. 
We have only on qc_link in the vector, which is `{.source_block_num=10, ,target_block_num=10, .is_link_strong=false}`

Reply:
I think that we may need to create a valid `ref` for the genesis block in `finality_core::create_core_for_genesis_block`, so maybe it should take a `block_state_ptr`, or `id + timestamp + finalizer_policy_generation`